src/explore-education-statistics-common/src/components/util/__tests__/generatePageNumbers.test.ts

```diff
@@ -0,0 +1,453 @@
+        currentPage: 8,
+        maxReturnSize: 12,
+        totalPages: 15,
+      }),
+    ).toEqual([1, null, 5, 6, 7, 8, 9, 10, 11, null, 15]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 9,
+        maxReturnSize: 12,
+        totalPages: 15,
+      }),
+    ).toEqual([1, null, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 10,
+        maxReturnSize: 12,
+        totalPages: 15,
+      }),
+    ).toEqual([1, null, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 11,
+        maxReturnSize: 12,
+        totalPages: 15,
+      }),
+    ).toEqual([1, null, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 12,
+        maxReturnSize: 12,
+        totalPages: 15,
+      }),
+    ).toEqual([1, null, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 13,
+        maxReturnSize: 12,
+        totalPages: 15,
+      }),
+    ).toEqual([1, null, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 14,
+        maxReturnSize: 12,
+        totalPages: 15,
+      }),
+    ).toEqual([1, null, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 15,
+        maxReturnSize: 12,
+        totalPages: 15,
+      }),
+    ).toEqual([1, null, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
+  });
+
+  test('uses the last page as the current page when currentPage is greater than totalPages', () => {
+    expect(
+      generatePageNumbers({
+        currentPage: 6,
+        maxReturnSize: 7,
+        totalPages: 4,
+      }),
+    ).toEqual([1, 2, 3, 4]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 12,
+        maxReturnSize: 7,
+        totalPages: 10,
+      }),
+    ).toEqual([1, null, 6, 7, 8, 9, 10]);
+  });
+
+  test('uses the first page as the current page when currentPage is 0', () => {
+    expect(
+      generatePageNumbers({
+        currentPage: 0,
+        maxReturnSize: 7,
+        totalPages: 4,
+      }),
+    ).toEqual([1, 2, 3, 4]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 0,
+        maxReturnSize: 7,
+        totalPages: 10,
+      }),
+    ).toEqual([1, 2, 3, 4, 5, null, 10]);
+  });
+
+  test('uses the minimum return size if maxReturnSize below it', () => {
+    expect(
+      generatePageNumbers({
+        currentPage: 1,
+        maxReturnSize: 3,
+        totalPages: 6,
+      }),
+    ).toEqual([1, 2, 3, null, 6]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 2,
+        maxReturnSize: 3,
+        totalPages: 6,
+      }),
+    ).toEqual([1, 2, 3, null, 6]);
+
+    expect(
+      generatePageNumbers({
+        currentPage: 3,
+        maxReturnSize: 3,
+        totalPages: 6,
+      }),
+    ).toEqual([1, 2, 3, null, 6]);
```

Comment:
That would take the size above the minimum max size (5), currently it will only return the max size or lower. If the numbers on either side of the current one don't fit into that then one is left out.
Not sure if that's necessarily right, but I haven't found much of a consensus when looking at pagination patterns.

Reply:
Ah, I made a little mistake in my comment - 1, 2, 3, 4, null, 6 would be what I would expect.

I think my main issue here is that we wouldn't be respecting the GDS pattern's expectations.

> For smaller screens, show page numbers for:
> 
> - the current page
> - previous and next pages
> - first and last pages
>
> For larger screens, show page numbers for:
> 
> - the current page
> - at least one page immediately before and after the current page
> - first and last pages
> 
> Use ellipses (…) to replace any skipped pages.

The main point is that prev/next numbers have to be covered so it's it's possible to have up to 7 items as the minimum (first, spacer, prev, current, next, spacer, last).

I'm kinda thinking that specifying a max number of items is causing confusion here and perhaps specifying a target window size might work better? A window size of 1 would be the bare minimum as this accounts for prev/next. Window sizes of 2+ would result in stuff like 1, null, 3, 4, 5, 6, 7, null, 10, etc.